File: src/Cartridge.cpp
#include "Cartridge.h"
#include "Mapper_000.h"
#include "Mapper_001.h"
#include "Mapper_002.h"
#include "Mapper_004.h"
#include "Mapper_005.h"
#include "Mapper_069.h"
#include <fstream>
#include <iostream>

Cartridge::Cartridge(const std::string &sFileName) {
  struct sHeader {
    char name[4];
    uint8_t prg_rom_chunks;
    uint8_t chr_rom_chunks;
    uint8_t mapper1;
    uint8_t mapper2;
    uint8_t prg_ram_size;
    uint8_t tv_system1;
    uint8_t tv_system2;
    char unused[5];
  } header;

  bImageValid = false;

  std::ifstream ifs(sFileName, std::ifstream::binary);
  if (ifs.is_open()) {
    ifs.read((char *)&header, sizeof(sHeader));

    if (header.mapper1 & 0x04)
      ifs.seekg(512, std::ios_base::cur);

    nMapperID = ((header.mapper2 >> 4) << 4) | (header.mapper1 >> 4);
    nPRGBanks = header.prg_rom_chunks;
    nCHRBanks = header.chr_rom_chunks;

    // Read hardware mirroring from ROM header (bit 0 of mapper1)
    MIRROR hwMirror =
        (header.mapper1 & 0x01) ? MIRROR::VERTICAL : MIRROR::HORIZONTAL;

    std::cout << "Mapper ID: " << (int)nMapperID << std::endl;
    std::cout << "PRG Banks: " << (int)nPRGBanks << std::endl;
    std::cout << "CHR Banks: " << (int)nCHRBanks << std::endl;
    std::cout << "Mirroring: "
              << (hwMirror == MIRROR::VERTICAL ? "Vertical" : "Horizontal")
              << std::endl;

    // Debug log to file
    std::ofstream debugLog("nes_debug.log", std::ios::app);
    debugLog << "Loading ROM - Mapper: " << (int)nMapperID
             << ", PRG: " << (int)nPRGBanks << ", CHR: " << (int)nCHRBanks
             << std::endl;
    debugLog.close();

    vPRGMemory.resize(nPRGBanks * 16384);
    ifs.read((char *)vPRGMemory.data(), vPRGMemory.size());

    if (nCHRBanks == 0) {
      // CHR RAM - allocate 8KB
      vCHRMemory.resize(8192);
    } else {
      vCHRMemory.resize(nCHRBanks * 8192);
      ifs.read((char *)vCHRMemory.data(), vCHRMemory.size());
    }

    switch (nMapperID) {
    case 0:
      pMapper = std::make_shared<Mapper_000>(nPRGBanks, nCHRBanks, hwMirror);
      break;
    case 1:
      pMapper = std::make_shared<Mapper_001>(nPRGBanks, nCHRBanks);
      break;
    case 2:
      pMapper = std::make_shared<Mapper_002>(nPRGBanks, nCHRBanks, hwMirror);
      break;
    case 4:
      pMapper = std::make_shared<Mapper_004>(nPRGBanks, nCHRBanks);
      break;
    case 5: {
      std::ofstream debugLog("nes_debug.log", std::ios::app);
      debugLog << "Creating Mapper 5 (MMC5)..." << std::endl;
      debugLog.close();
    }
      pMapper = std::make_shared<Mapper_005>(nPRGBanks, nCHRBanks);
      {
        std::ofstream debugLog("nes_debug.log", std::ios::app);
        debugLog << "Mapper 5 created successfully" << std::endl;
        debugLog.close();
      }
      break;
    case 69:
      pMapper = std::make_shared<Mapper_069>(nPRGBanks, nCHRBanks);
      {
        std::ofstream debugLog("nes_debug.log", std::ios::app);
        debugLog << "Mapper 69 (Sunsoft FME-7) created successfully"
                 << std::endl;
        debugLog.close();
      }
      break;
    default:
      std::cerr << "Unsupported Mapper: " << (int)nMapperID << std::endl;
      {
        std::ofstream debugLog("nes_debug.log", std::ios::app);
        debugLog << "UNSUPPORTED Mapper: " << (int)nMapperID << std::endl;
        debugLog.close();
      }
      pMapper = nullptr;
      bImageValid = false;
      ifs.close();
      return;
    }

    bImageValid = true;
    ifs.close();
  }
}

Cartridge::~Cartridge() {}

void Cartridge::reset() {
  if (pMapper)
    pMapper->reset();
}

bool Cartridge::ImageValid() { return bImageValid; }

bool Cartridge::cpuRead(uint16_t addr, uint8_t &data) {
  uint32_t mapped_addr = 0;
  if (pMapper && pMapper->cpuMapRead(addr, mapped_addr)) {
    if (mapped_addr == 0xFFFFFFFF) {
      // PRG RAM or register access
      Mapper_001 *m1 = dynamic_cast<Mapper_001 *>(pMapper.get());
      if (m1) {
        data = m1->GetPRGRAM()[addr & 0x1FFF];
        return true;
      }
      Mapper_004 *m4 = dynamic_cast<Mapper_004 *>(pMapper.get());
      if (m4) {
        data = m4->GetPRGRAM()[addr & 0x1FFF];
        return true;
      }
      Mapper_005 *m5 = dynamic_cast<Mapper_005 *>(pMapper.get());
      if (m5) {
        if (addr >= 0x5000 && addr <= 0x5FFF) {
          // MMC5 register read
          data = m5->ReadRegister(addr);
        } else if (addr >= 0x6000 && addr <= 0x7FFF) {
          // PRG RAM at $6000-$7FFF
          data = m5->GetPRGRAM()[addr & 0x1FFF];
        } else {
          // PRG RAM mapped to $8000+ region
          data = m5->GetPRGRAM()[addr & 0x1FFF];
        }
        return true;
      }
      Mapper_069 *m69 = dynamic_cast<Mapper_069 *>(pMapper.get());
      if (m69) {
        data = m69->GetPRGRAM()[addr & 0x1FFF];
        return true;
      }
      return true;
    }
    if (mapped_addr < vPRGMemory.size()) {
      data = vPRGMemory[mapped_addr];
    }
    return true;
  }
  return false;
}

bool Cartridge::cpuWrite(uint16_t addr, uint8_t data) {
  uint32_t mapped_addr = 0;
  if (pMapper && pMapper->cpuMapWrite(addr, mapped_addr, data)) {
    if (mapped_addr == 0xFFFFFFFF) {
      // PRG RAM handled inside mapper
      return true;
    }
    if (mapped_addr < vPRGMemory.size()) {
      vPRGMemory[mapped_addr] = data;
    }
    return true;
  }
  return false;
}

bool Cartridge::ppuRead(uint16_t addr, uint8_t &data) {
  uint32_t mapped_addr = 0;
  if (pMapper->ppuReadCustom(addr, data)) {
    return true;
  }
  if (pMapper && pMapper->ppuMapRead(addr, mapped_addr)) {
    if (mapped_addr < vCHRMemory.size()) {
      data = vCHRMemory[mapped_addr];
    }
    return true;
  }
  return false;
}

bool Cartridge::ppuWrite(uint16_t addr, uint8_t data) {
  uint32_t mapped_addr = 0;
  if (pMapper->ppuWriteCustom(addr, data)) {
    return true;
  }
  if (pMapper && pMapper->ppuMapWrite(addr, mapped_addr)) {
    if (mapped_addr < vCHRMemory.size()) {
      vCHRMemory[mapped_addr] = data;
    }
    return true;
  }
  return false;
}

File: src/Mogiki.Core/Cartridge/Cartridge.cs
using Mogiki.Core.Common;
using Mogiki.Core.Mappers;

namespace Mogiki.Core.Cartridge;

/// <summary>
/// Represents an iNES ROM cartridge and its mapper hardware.
/// </summary>
public sealed class Cartridge
{
    public bool ImageValid { get; private set; }
    public byte MapperId { get; private set; }
    public byte PrgBanks { get; private set; }
    public byte ChrBanks { get; private set; }

    public byte[] PrgMemory { get; private set; } = [];
    public byte[] ChrMemory { get; private set; } = [];
    public byte[] PrgRam { get; } = new byte[8192];

    public Mapper? Mapper { get; private set; }

    public Cartridge(string fileName)
    {
        if (!File.Exists(fileName))
        {
            ImageValid = false;
            return;
        }

        using var stream = File.OpenRead(fileName);
        using var reader = new BinaryReader(stream);

        if (stream.Length < 16)
        {
            ImageValid = false;
            return;
        }

        byte[] headerBytes = reader.ReadBytes(16);
        if (headerBytes[0] != 'N' || headerBytes[1] != 'E' || headerBytes[2] != 'S' || headerBytes[3] != 0x1A)
        {
            ImageValid = false;
            return;
        }

        PrgBanks = headerBytes[4];
        ChrBanks = headerBytes[5];
        byte mapper1 = headerBytes[6];
        byte mapper2 = headerBytes[7];

        // Trainer (512 bytes)
        if ((mapper1 & 0x04) != 0)
        {
            reader.ReadBytes(512);
        }

        MapperId = (byte)(((mapper2 >> 4) << 4) | (mapper1 >> 4));
        var hwMirror = (mapper1 & 0x01) != 0 ? MirrorMode.Vertical : MirrorMode.Horizontal;

        PrgMemory = reader.ReadBytes(PrgBanks * 16384);

        if (ChrBanks == 0)
        {
            // CHR RAM - allocate 8KB
            ChrMemory = new byte[8192];
        }
        else
        {
            ChrMemory = reader.ReadBytes(ChrBanks * 8192);
        }

        Mapper = MapperId switch
        {
            0 => new Mapper000(PrgBanks, ChrBanks, hwMirror),
            1 => new Mapper001(PrgBanks, ChrBanks),
            2 => new Mapper002(PrgBanks, ChrBanks, hwMirror),
            3 => new Mapper003(PrgBanks, ChrBanks, hwMirror),
            4 => new Mapper004(PrgBanks, ChrBanks),
            5 => new Mapper005(PrgBanks, ChrBanks),
            69 => new Mapper069(PrgBanks, ChrBanks),
            _ => null
        };

        ImageValid = Mapper != null;
    }

    public Cartridge(byte[] prgRom, byte[] chrRom, byte mapperId, MirrorMode hwMirror = MirrorMode.Horizontal)
    {
        PrgBanks = (byte)(prgRom.Length / 16384);
        ChrBanks = (byte)(chrRom.Length / 8192);
        MapperId = mapperId;
        PrgMemory = prgRom;
        ChrMemory = chrRom.Length == 0 ? new byte[8192] : chrRom;

        Mapper = MapperId switch
        {
            0 => new Mapper000(PrgBanks, ChrBanks, hwMirror),
            1 => new Mapper001(PrgBanks, ChrBanks),
            2 => new Mapper002(PrgBanks, ChrBanks, hwMirror),
            3 => new Mapper003(PrgBanks, ChrBanks, hwMirror),
            4 => new Mapper004(PrgBanks, ChrBanks),
            5 => new Mapper005(PrgBanks, ChrBanks),
            69 => new Mapper069(PrgBanks, ChrBanks),
            _ => null
        };

        ImageValid = Mapper != null;
    }

    public void Reset()
    {
        Mapper?.Reset();
    }

    public MirrorMode Mirror => Mapper?.Mirror ?? MirrorMode.Horizontal;
    public bool IrqState => Mapper?.IrqState ?? false;
    public void ClearIrq() => Mapper?.IrqClear();
    public void Scanline() => Mapper?.Scanline();
    public void Scanline(int currentScanline, int currentCycle) => Mapper?.Scanline(currentScanline, currentCycle);
    public void CpuSnoopWrite(ushort addr, byte data) => Mapper?.CpuSnoopWrite(addr, data);
    public void CpuSnoopRead(ushort addr) => Mapper?.CpuSnoopRead(addr);

    public bool CpuRead(ushort addr, out byte data)
    {
        data = 0;
        if (Mapper == null) return false;

        if (Mapper.CpuMapRead(addr, out uint mappedAddr))
        {
            if (mappedAddr == Mapper.PrgRamSentinel)
            {
                if (Mapper is Mapper001 m1)
                {
                    data = m1.PrgRam[addr & 0x1FFF];
                    return true;
                }
                if (Mapper is Mapper004 m4)
                {
                    data = m4.PrgRam[addr & 0x1FFF];
                    return true;
                }
                if (Mapper is Mapper005 m5)
                {
                    if (addr is >= 0x5000 and <= 0x5FFF)
                        data = m5.ReadRegister(addr);
                    else
                        data = m5.ReadPrgRam(addr);
                    return true;
                }
                if (Mapper is Mapper069 m69)
                {
                    data = m69.PrgRam[addr & 0x1FFF];
                    return true;
                }
                return true;
            }

            if (mappedAddr < PrgMemory.Length)
            {
                data = PrgMemory[mappedAddr];
                return true;
            }
            return true;
        }
        if (addr is >= 0x6000 and <= 0x7FFF)
        {
            data = PrgRam[addr & 0x1FFF];
            return true;
        }

        return false;
    }

    public bool CpuWrite(ushort addr, byte data)
    {
        if (Mapper == null) return false;

        if (Mapper.CpuMapWrite(addr, out uint mappedAddr, data))
        {
            if (mappedAddr == Mapper.PrgRamSentinel)
            {
                return true;
            }
            if (mappedAddr < PrgMemory.Length)
            {
                PrgMemory[mappedAddr] = data;
                return true;
            }
            return true;
        }
        if (addr is >= 0x6000 and <= 0x7FFF)
        {
            PrgRam[addr & 0x1FFF] = data;
            return true;
        }

        return false;
    }

    public bool PpuRead(ushort addr, out byte data)
    {
        data = 0;
        if (Mapper == null) return false;

        if (Mapper.PpuReadCustom(addr, out data))
        {
            return true;
        }

        if (Mapper.PpuMapRead(addr, out uint mappedAddr))
        {
            if (mappedAddr < ChrMemory.Length)
            {
                data = ChrMemory[mappedAddr];
                return true;
            }
            return true;
        }
        return false;
    }

    public bool PpuWrite(ushort addr, byte data)
    {
        if (Mapper == null) return false;

        if (Mapper.PpuWriteCustom(addr, data))
        {
            return true;
        }

        if (Mapper.PpuMapWrite(addr, out uint mappedAddr))
        {
            if (mappedAddr < ChrMemory.Length)
            {
                ChrMemory[mappedAddr] = data;
                return true;
            }
            return true;
        }
        return false;
    }
}
